Specify aquiring import maps

hiroshige-g opened the pull request:
This PR the semantics around "aquiring import maps", including
how import maps are loaded via <script> elements, and
how import map loading and module script loading interact and block each other.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/hiroshige-g/import-maps/pull/136.html" title="Last updated on Jun 24, 2019, 4:16 AM UTC (a45cf9b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/import-maps/136/cbd5e7d...hiroshige-g:a45cf9b.html" title="Last updated on Jun 24, 2019, 4:16 AM UTC (a45cf9b)">Diff</a>

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+        1. Increment |settings object|'s
+           [=environment settings object/pending import maps count=].
+        1. Fetch an import map given url, |settings object|, and options.
+           <p class="note">To be specced. This returns a string or a failure.</p>
```
Done. I'm planning to land this PR with "merge import maps" and "fetch an import map" definitions open.

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+               an event named error at the element, and return.
+            1. <a>Register an import map</a> given the result,
+               |settings object|, |base URL|, and the element.
+            1. Decrement |settings object|'s
```
Good catch! Moved this step before early return on failure.

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+        1. Return.
+- Insert the following before <a spec="html">prepare a script</a> Step 25.2:
+    - If <a spec="html">the script's type</a> is "importmap", then:
+        1. <a spec="html">Immediately</a> <a>register an import map</a>
```
Removed. I copied "Immediately" from Step 26 of https://html.spec.whatwg.org/multipage/scripting.html#prepare-a-script and I don't know its historical context.

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+be cleared for http://:invalid, because we early-exit module loading.
+</p>
+
+<h3 id="integration-register-an-import-map">Registering an import map</h3>
```
This section is mainly about modifying settings object's import maps, not fetching scripts.
Fetching scripts are directly affected/modified by changes to resolve-a-module-specifier.
WDYT?

domenic replied:
I think you are right, this is separate from fetching scripts. It is more like "Calling scripts", but import maps are different enough they deserve their own section. Let's leave it as-is.

hiroshige-g commented:
Thank you for review! I reflected some of comments, and will address the rest later.

> Using two-space indent
> Not doing any manual line wrapping (instead using your text editor's line wrapping)?

Is there any formatting tool (e.g. something clang-format-like) for bikeshed?

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+
+<p class="note">Alternatives: We might want to associate import maps with e.g.
+[=environment settings object/module map=], realm, etc.
+So far I don't have any specific pros/cons for these options.</p>
```
Sounds good. Do you think we also move `pending import maps count` and `acquiring import maps flag` to Document/WorkerGlobalScope etc. to keep environment settings object purely abstract?

hiroshige-g replied:
Moved `import map` to Document and WorkerGlobalScope.

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+run these steps.
+
+1. Block while |settings object|'s
+   [=environment settings object/pending import maps count=] is not 0.
```
Done.

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+1. <p class="issue">Should we check whether the script element is moved
+   across documents?
+   Currently, we don't check that and the import maps are always registered
+   to the settings object at the time of prepare-a-script.</p>
```
I added a check for moving between documents, and add a link to the spec PR https://github.com/whatwg/html/pull/2673 for other script cases.
I feel this PR is not strongly blocked by #2673, as we can amend the the condition here later according to the future discussion on #2673.

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+       and |import map|.
+
+To <dfn>merge import maps</dfn>, given an [=/import map=] |import map| and
+an [=/import map=] |new import map|:
```
nit: for me, "concatenate" sounds like non-in-place (as JavaScript's array's concat() method doesn't modify the original in-place), but this is not a strong preference.

hiroshige-g replied:
Renamed to "update import maps".

domenic commented on spec.bs:
```diff
@@ -59,16 +60,250 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+Each [=environment settings object=] will get an <dfn for="environment settings object">import map</dfn> algorithm, which returns an [=/import map=] created by parsing and merging all `<script type="importmap">` elements that are encountered (before the cutoff).
+
+A <a spec="html">Document</a> has an [=/import map=] <dfn for="Document">import map</dfn>.
+It is initially null.
```
Shouldn't it be an empty import map? Something like "it is initially an import map with an imports and scopes both being empty maps".

hiroshige-g replied:
Done. Defined "empty import map".

domenic commented on spec.bs:
```diff
@@ -59,16 +60,250 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+
+Each [=environment settings object=] will get an <dfn for="environment settings object">import map</dfn> algorithm, which returns an [=/import map=] created by parsing and merging all `<script type="importmap">` elements that are encountered (before the cutoff).
+
+A <a spec="html">Document</a> has an [=/import map=] <dfn for="Document">import map</dfn>.
```
Use `{{Document}}`. This may require adding a link-defaults

hiroshige-g replied:
Done.

domenic commented on spec.bs:
```diff
@@ -59,16 +60,250 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+<var ignore>settings object</var>'s [=environment settings object/import map=] returns
+the [=Document/import map=] of <var ignore>window</var>'s <a spec="html">associated Document</a>.
+
+A <dfn>WorkerGlobalScope</dfn> has an [=/import map=] <dfn for="WorkerGlobalScope">import map</dfn>.
```
Try using `{{WorkerGlobalScope}}`

hiroshige-g replied:
Thanks! Done.

domenic commented on spec.bs:
```diff
@@ -59,16 +60,250 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+run the following steps.
+1. If |source text| is a failure:
+    1. <a spec="html">queue a task</a> to <a spec="html">fire an event</a> named `error` at
+       |element|, and return.
```
Nit: this algorithm has a lot of single-step nested lists. Instead of `1. If x, \n  1. y` just say `If x, then y`, on one line.

hiroshige-g replied:
Done.

hiroshige-g commented on spec.bs:
```diff
@@ -59,16 +60,200 @@ A <dfn>import map</dfn> is a [=struct=] with two [=struct/items=]:
+       and |import map|.
+
+To <dfn>merge import maps</dfn>, given an [=/import map=] |import map| and
+an [=/import map=] |new import map|:
```
Done.

hiroshige-g commented:
Updated the PR.

I tried to move `pending import maps count` and `acquiring import maps flag` to Document, but I couldn't make the environment settings object's members read-only, because `wait for import maps` have to modify the `acquiring import maps flag` via `settings object`. I left these flags as-is. Any opinions?

Also fixed wait for import maps (6e71eb2).

Remaining TODO is to fix links to HTML specs (e.g. `fetch a module script graph` -> `fetch an external module script graph`) after the HTML-spec PR lands.

domenic commented:
> I tried to move pending import maps count and acquiring import maps flag to Document, but I couldn't make the environment settings object's members read-only, because wait for import maps have to modify the acquiring import maps flag via settings object. I left these flags as-is. Any opinions?

I will try moving this all to Document, since this acquiring stuff is Document-specific, i.e. we will probably use a different mechanism in workers.

domenic commented:
Moving it to Document is too hard because the module script fetching algorithms are all agnostic to document vs. worker. I guess it is OK as-is for now, although it is a bit strange to put state directly on environment settings objects. We can try to figure out a better solution later perhaps.

I pushed minor editorial fixes, but it mostly looks good to me now!

The only remaining issue I have is that I do not understand the second warning, that says

> Depending on the exact location of wait for import maps, there would be observable behavior differences between HTML-spec- and Fetch-spec-based import maps. For example, the current HTML-spec-based draft here calls wait for import maps at very early stages of module loading before resolve a module specifier. Therefore, acquiring import maps is always set back to false, even if resolve a module specifier fails e.g. in parsing http://:invalid. When we switch to Fetch-spec-based, import map resolution would be done inside Fetch spec after resolve a module specifier and thus we might also want to wait for import maps after resolve a module specifier. In that case, acquiring import maps wouldn’t be reset to false for http://:invalid, because we early-exit module loading. 

Can you expand (just on GitHub is fine) with an example? I'm not sure, for example, where `http://:invalid` appears. It's also unclear to me why we could reset to false with one design, but not the other---what prevents us from resetting to false even in a Fetch-spec-based design?

I'd like to get that cleared up before merging since I don't want to have text in the spec I don't understand :).

hiroshige-g commented:
> Can you expand (just on GitHub is fine) with an example? I'm not sure, for example, where http://:invalid appears. It's also unclear to me why we could reset to false with one design, but not the other---what prevents us from resetting to false even in a Fetch-spec-based design?

I digged more into this and found that:

- `<script src="http://:invalid" type="module">`
  - NOT reset to false (because of early return at Step 24.5 of https://html.spec.whatwg.org/multipage/scripting.html#prepare-a-script).
- `new Worker("http://:invalid", {type: "module"})`
  - NOT reset to false (because of early return at Step 4 of https://html.spec.whatwg.org/multipage/workers.html#dom-worker).
- `import('http://:invalid');`
  - RESET to false in the current PR, by the `wait for import maps` call at the head of `fetch an import() module script graph`.
    - Avoiding resetting the flag requires additional complexity inside resolve-a-module-specifier.
  - NOT reset to false if we move `wait for import maps` to Fetch spec.
    - I basically assume that, as for the resolution part (not the registering part) "Fetch-spec-based" import maps modifies Fetch spec only (e.g. `wait for import maps` calls are inserted to Fetch spec), to support many non-module script cases with a single modification to Fetch spec, without adding many modifications for each resource type to HTML spec.
    - If we keep some logic at the head of `fetch an import() module script graph` to reset the flag, we'll still have "RESET to false" behavior.

So `import('http://:invalid');`'s flag resetting behavior

- Might be changed when switching to Fetch spec,
- But it looks a little inconsistent with `<script src="http://:invalid" type="module">` even within the current PR.

Probably it's good to land this PR with the second warning removed, and file a github issue about this.

domenic commented:
Thanks for explaining! I will reword the note a bit. I don't think it's a big deal to have this inconsistency because import() is so different (operating on specifiers instead of URLs).
